C++/include/algorithm/dynamic_programming/README.md

```diff
@@ -44,3 +50,197 @@ Time               | Space
+
+### Complexity
+
+Time               | Space
+-------------------|-------------------
+_O(M*N)_ | _O(M*N)_
+
+where M is the number of coins with different values and 
+N is the amount that we desired to change.
+
+## 3. 0_1 Knapsack
+
+Given weights and values of n items, put these items in a knapsack of a fixed capacity to get the maximum total value in the knapsack.
+
+### Input
+- #### `capacity`
+  Type: `long long`
+
+  Represents the maximum capacity of the knapsack, in which we have to put weights.
+
+- #### `numberOfItems`
+  Type: `long long`
+
+  Represents the total number of items given.
+
+- #### `weights`
+  Type: `vector<long long>`
+
+  vector of the weights for the items given.
+
+- #### `values`
+  Type: `vector<long long>`
+
+  vector of the values for the items given.
+
+### Output
+- #### `maxItems[numberOfItems][capacity]`
+  Type: `int (bottom rightmost cell of 2D Matrix of type vector<vector<long long>>)`
+
+  After filling the matrix the bottom rightmost cell of the of the matrix will contain maximum total value in the knapsack.  
+
+### Complexity
+
+Time               | Space
+-------------------|-------------------
+_O(N*C)_ | _O(N*C)_
+
+where N is the number of items and C is the capacity of the knapsack.
+
+## 4. Kadane
+
+This algorithm is used for finding the contiguous subarray within a one-dimensional array of integers which has the largest sum.
+
+### Input
+- #### `values`
+  Type: `vector<int>`
+
+  Vector of integers for which we have to find the contiguous subarray having the largest sum.
+
+### Output
+- #### `maxSum, start, end`
+  Type: `tuple<int, size_t, size_t>`
+
+  Output is a tuple of maximum sum, start index of the contiguous array and the last index of the contiguous array. Note: `size_t` is an unsigned integral data type.  
+
+### Complexity
+
+Time               | Space
+-------------------|-------------------
+_O(N)_ | _O(1)_
+
+where N is the number of elements in the original array.
+
+## 5. Longest Decreasing Subsequence
+
+This algorithm finds the longest decreasing subsequence in a sequence of numbers.
+
+### Input
+- #### `input`
+  Type: `vector<int>`
+
+  Vector array of inputs for which we have to find longest decreasing subsequence.
+
+- #### `to_show_state`
+  Type: `const bool`
+
+  Initially it is set to false, if no second argument is provided in the function. if this is true subsequence is printed out by the function.
+
+### Output
+- #### `max`
+  Type: `int`
+
+  Represents the largest length of the longest decreasing subsequence.  
+
+### Complexity
+
+Time               | Space
+-------------------|-------------------
+_O(n log(n))_ | _O(n)_
+
+where n is the number of elements in the initial list.
+
+## 6. Rod Cutting
+
+Given a rod of length N units and an array of prices that contains prices of all pieces of size smaller than N. Determine the maximum value obtainable by cutting up the rod and selling the pieces.
```

Comment:
"Given a rod of length N units and an array of prices that contains prices of all pieces of size smaller than N, determine the maximum value obtainable by cutting up the rod and selling the pieces."